pkg/controllers/subnet/subnet_controller.go

```diff
@@ -266,87 +303,100 @@ func deleteSuccess(r *SubnetReconciler, _ context.Context, o *v1alpha1.Subnet) {
 			}).
 		Complete(r)
 }
 
-func StartSubnetController(mgr ctrl.Manager, subnetService *subnet.SubnetService, subnetPortService servicecommon.SubnetPortServiceProvider, vpcService servicecommon.VPCServiceProvider) error {
-	subnetReconciler := &SubnetReconciler{
-		Client:            mgr.GetClient(),
-		Scheme:            mgr.GetScheme(),
-		SubnetService:     subnetService,
-		SubnetPortService: subnetPortService,
-		VPCService:        vpcService,
-		Recorder:          mgr.GetEventRecorderFor("subnet-controller"),
-	}
-	if err := subnetReconciler.Start(mgr); err != nil {
-		log.Error(err, "failed to create controller", "controller", "Subnet")
-		return err
+func (r *SubnetReconciler) listSubnetCRs(ctx context.Context) ([]string, error) {
+	crdSubnetList := &v1alpha1.SubnetList{}
+	err := r.Client.List(ctx, crdSubnetList)
+	if err != nil {
+		return nil, err
 	}
-	go common.GenericGarbageCollector(make(chan bool), servicecommon.GCInterval, subnetReconciler.CollectGarbage)
-	return nil
-}
 
-// Start setup manager
-func (r *SubnetReconciler) Start(mgr ctrl.Manager) error {
-	err := r.setupWithManager(mgr)
-	if err != nil {
-		return err
+	crdSubnetIDs := make([]string, 0, len(crdSubnetList.Items))
+	for _, sr := range crdSubnetList.Items {
+		crdSubnetIDs = append(crdSubnetIDs, string(sr.UID))
 	}
-	return nil
+	return crdSubnetIDs, nil
 }
 
-// CollectGarbage implements the interface GarbageCollector method.
-func (r *SubnetReconciler) CollectGarbage(ctx context.Context) {
-	log.Info("subnet garbage collector started")
-	crdSubnetList := &v1alpha1.SubnetList{}
-	err := r.Client.List(ctx, crdSubnetList)
+// collectGarbage implements the interface GarbageCollector method.
+func (r *SubnetReconciler) collectGarbage(ctx context.Context) {
+	startTime := time.Now()
+	defer func() {
+		log.Info("Subnet garbage collection completed", "time", time.Since(startTime))
+	}()
+
+	crdSubnetIDs, err := r.listSubnetCRs(ctx)
 	if err != nil {
-		log.Error(err, "failed to list subnet CR")
+		log.Error(err, "failed to list Subnet CRs")
 		return
 	}
+
 	var nsxSubnetList []*model.VpcSubnet
-	for _, subnet := range crdSubnetList.Items {
-		nsxSubnetList = append(nsxSubnetList, r.SubnetService.ListSubnetCreatedBySubnet(string(subnet.UID))...)
+	for _, crdSubnetID := range crdSubnetIDs {
+		nsxSubnetList = append(nsxSubnetList, r.SubnetService.ListSubnetCreatedBySubnet(crdSubnetID)...)
 	}
 	if len(nsxSubnetList) == 0 {
+		log.Info("No Subnets found in NSX, garbage collection complete")
 		return
 	}
 
-	crdSubnetIDs := sets.NewString()
-	for _, sr := range crdSubnetList.Items {
-		crdSubnetIDs.Insert(string(sr.UID))
-	}
-
-	for _, elem := range nsxSubnetList {
-		uid := util.FindTag(elem.Tags, servicecommon.TagScopeSubnetCRUID)
-		if crdSubnetIDs.Has(uid) {
+	crdSubnetIDsSet := sets.NewString(crdSubnetIDs...)
+	for _, nsxSubnet := range nsxSubnetList {
+		uid := nsxutil.FindTag(nsxSubnet.Tags, servicecommon.TagScopeSubnetCRUID)
+		if crdSubnetIDsSet.Has(uid) {
```

Comment:
The logic is changed here, you need to try to delete NSX Subnet event the CR is existing.
Subnet GC should two things:
1) scale down NSX Subnet if there is no port attached on it, this is current GC doing.
2) delete stale NSX Subnet if there is no Subnet CR existing

we can have a following patch or this patch for 2), but please also keep 1) logic in GC.